Ensure DiscardNode() is faithfully called

edmc-ss opened the pull request:
Previously, there were a number of cases where DiscardNode() would
not be called. Most particularly, during DeleteBy*() rebalancing,
it is possible that two nodes are merged and/or the height of the
tree is reduced. In these cases, the node going away was not marked
as being discarded and would then never be "free'd" from the backing
store for the B+Tree.

The other big opportunity for leakage is when a B+Tree is being
disposed of via a call to Discard(). Any nodes of the B+Tree not
currently in memory would not be scheduled for calls to DiscardNode().

Commit:
Ensure DiscardNode() is faithfully called

Previously, there were a number of cases where DiscardNode() would
not be called. Most particularly, during DeleteBy*() rebalancing,
it is possible that two nodes are merged and/or the height of the
tree is reduced. In these cases, the node going away was not marked
as being discarded and would then never be "free'd" from the backing
store for the B+Tree.

The other big opportunity for leakage is when a B+Tree is being
disposed of via a call to Discard(). Any nodes of the B+Tree not
currently in memory would not be scheduled for calls to DiscardNode().

edmc-ss commented:
Admittedly, I did not pull over the ASSERTing code changes prototyped by @charmster elsewhere. As we discussed in that review, there remains an open question about what to do when such ASSERTs fail. As such, there were also some changes that more or less ensured that a node being discarded was, in fact, "empty". In order for that ASSERT code to not flag, we'd also need to port over the changes that make the node being discarded empty.

charmster commented on btree.go:
```diff
@@ -994,62 +950,43 @@ func (tree *btreeTreeStruct) Discard() (err error) {
-	err = nil
-	return
-}
+		for nil != staleOnDiskReferenceElement {
```
this Front()/Remove() loop might be better written as:
```
	if nil != tree.staleOnDiskReferencesList {
            for staleOnDiskReferenceElement = tree.staleOnDiskReferencesList.Front();
                staleOnDiskReferenceElement != nil;
                tree.staleOnDiskReferencesList.Remove(staleOnDiskReferenceElement) {

                 ... call DiscardNode() for the backing object ...
            }
        }
```

edmc-ss replied:
I switched to iterating over a map (a "set" using a map who's values are the minimally sized struct{}).

charmster commented on btree.go:
```diff
@@ -2019,7 +1952,112 @@ func (tree *btreeTreeStruct) markNodeEvicted(node *btreeNodeStruct) {
+		tree.nodeCache.Lock()
+		switch node.btreeNodeCacheTag {
+		case noLRU:
+			err := fmt.Errorf("Logic error in markNodeEvicted() with node.btreeNodeCacheTag == noLRU (%v)", noLRU)
```
this message should probably be markNodeToBeDiscarded() instead of markNodeEvicted().

edmc-ss replied:
Cut&Paste gaff - thx

charmster commented on btree.go:
```diff
@@ -2019,7 +1952,112 @@ func (tree *btreeTreeStruct) markNodeEvicted(node *btreeNodeStruct) {
 		switch node.btreeNodeCacheTag {
 		case noLRU:
-			err := fmt.Errorf("Logic error in markNodeUsed() with node.btreeNodeCacheTag == noLRU (%v)", noLRU)
+			err := fmt.Errorf("Logic error in markNodeEvicted() with node.btreeNodeCacheTag == noLRU (%v)", noLRU)
```
its looks like we're adding a 4th copy of the noLRU/cleanLRU/dirtyLRU list maintenance code.  lovely.

if the doubly-linked lists were implemented differently this could be simpler and cleaner such that the subtle differences between the sundry markNode*() routines be more obvious.

edmc-ss replied:
Totally admit cut&paste ugliness...

charmster reviewed (approved):
It looks like this fixes the same two bugs that my changes fixed.  i don't think it fixes any additional bugs. 
 the markNodeToBeDiscarded() that you added in the
```
    // height will reduce by one, so make leftSiblingNode the new root
```
cases was already fixed in the original code because the parent node is always marked dirty.

these changes don't remove the DiscardNode() logic from nodeMarkDirty() so i'm not sure what the point of it all is.  in some cases where you call nodeMarkToBeDiscarded() the node has already been passed to nodeMarkDirty() so the "queue for Discard" logic happened at that point.

actually, this may fix one potential bug:  if a client of sortedmap were to modify a tree and then, instead of calling tree.Flush() it called tree.Discard() then we would fail to run tree.Prune() to handle the nodes were sitting on the list.  but current proxyfs code doesn't do that (except, perhaps, in some error paths that likely leave the file system corrupted as well).

Interestingly, now that i look at flushNode() after our conversation on Friday(?) i see that flushNode() walks the entire tree -- it doesn't stop when it hits a node that's not dirty (though it does stop when it finds a node that's not loaded), so the check i added there that the parent node must be dirty if a child is dirty would have caught a case where the parent wasn't dirty.

as i mentioned before, i'd like to preserve, in some sense, the tracing and Asserts that i added to the code so i'm going to merge my pull request to the Proxyfs "vendor" version of the code and then you can over-write it when you do the "glide up" thing.
￼

edmc-ss commented:
So the call to nodeMarkToBeDiscarded() for the disappearing root node is importantly different than the nodeMarkDirty() you mention as having been called... the node is *removed from the LRUs* in the former... and placed in the Dirty LRU in the latter. Very key!

But your comment *did* trigger an investigation I need to do - the whole reason I now recall that I used a sortedmap over a list.List was that it is important an element isn't on the list twice. Of course a map would work for this as well - and I think I'll promptly replace the list.List with a map. Although order is not guaranteed, we can fortunately iterate over a map.

But to your point about markNodeDirty() having been called for it - that would make sense we certainly have populated the dirty state up to and including the root by this point.

So let me address the potential "double free" bug I just introduced and re-push :-)

Commit:
Convert staleOnDiskReferencesList to use an underlying map

Previously, the use of a list.List would allow a discarded B+Tree node
to be free'd twice if, say, it happened to be first marked "dirty" and
later marked "to be discarded". This change takes advantage of the fact
that structs may be used as keys to a map (if they are used "by value"
instead of "by reference") and that duplicate modifications of a key
are collapsed if the value is empty since the 2nd write overwrites the
first. Here, we are using a struct{} as a value to get Go's map to
emulate a set. Finally, while iterating over a map does not guarantee
the order of returned keys, it is sufficient for the purposes of a list
use case where all elements will be "consumed".

edmc-ss reviewed (commented):
I *still* have not addressed @charmster's desire to have logging in here... nor asserts (with panics) alas.
